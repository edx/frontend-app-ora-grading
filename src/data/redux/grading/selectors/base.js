import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import * as module from './base';

export const rootSelector = ({ grading }) => grading;
export const mkSimpleSelector = (key) => createSelector(
  [module.rootSelector],
  (grading) => grading[key],
);
export const rootKeys = StrictDict({
  selected: 'selected',
  activeIndex: 'activeIndex',
  current: 'current',
  gradeData: 'gradeData',
  gradingData: 'gradingData',
});

export const simpleSelectors = {
  selected: mkSimpleSelector(rootKeys.selected),
  activeIndex: mkSimpleSelector(rootKeys.activeIndex),
  current: mkSimpleSelector(rootKeys.current),
  gradeData: mkSimpleSelector(rootKeys.gradeData),
  gradingData: mkSimpleSelector(rootKeys.gradingData),
};

/**
 * returns the length of the list of selected submissions
 * @return {number} selected submission list length
 */
export const selectionLength = createSelector(
  [module.simpleSelectors.selected],
  (selected) => selected.length,
);

export default {
  ...simpleSelectors,
  selectionLength,
};
