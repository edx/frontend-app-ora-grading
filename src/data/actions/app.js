import { StrictDict } from 'utils';
// import { createActionFactory } from './utils';

// export const dataKey = 'app';
// const createAction = createActionFactory(dataKey);

export default StrictDict({ });
