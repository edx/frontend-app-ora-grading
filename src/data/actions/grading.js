import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'grading';
const createAction = createActionFactory(dataKey);

/**
 * Load the first of the selected submission list for review, and initializes
 * the review pane to the first index.
 * @param {obj} submission data for the review/grading view
 *   {
 *     {obj} response - api response data
 *     {obj} gradeData - api grade data
 *     {str} status - api grade status
 *   }
 */
const loadSubmission = createAction('loadSubmission');

/**
 * Pre-load just the static info about the "next" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadNext = createAction('preloadNext');

/**
 * Pre-load just the static info about the "previous" submission in the review queue.
 * Load submission and the learner's response.
 * @param {obj} submission ({ response })
 */
const preloadPrev = createAction('preloadPrev');

/**
 * Load the "next" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadNext = createAction('loadNext');

/**
 * Load the "prev" submission in the selected queue as the current selection, load its current
 * status and grade data, and update prev/next accordingly.
 * @param {obj} { status, gradeData }
 */
const loadPrev = createAction('loadPrev');

/**
 * Load the selected submissions, storing their static data in an ordered array and setting the starting
 * index at the beginning of the list.
 * @param {obj[]} selection - ordered array of submission static data for all selected submissions
 */
const updateSelection = createAction('updateSelection');

// TODO: implement/design data workflow
const rubric = StrictDict({
  /*
   * update the local version of the rubric-level comment
   * @param {string} comment
   */
  updateComment: createAction('rubric/comment'),
  /*
   * update the local version of points for the given criterion
   * @param {number} index
   * @param {number} points
   */
  updateCriterionPoints: createAction('rubric/criterionPoints'),
  /*
   * update the local version of comment for the given criterion
   * @param {number} index
   * @param {string} comments
   */
  updateCriterionComment: createAction('rubric/criterionComment'),
});

export default StrictDict({
  loadSubmission,
  preloadNext,
  preloadPrev,
  loadNext,
  loadPrev,
  updateSelection,
  rubric,
});
