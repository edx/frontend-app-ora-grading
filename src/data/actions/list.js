import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'list';
const createAction = createActionFactory(dataKey);

export default StrictDict({
  load: createAction('load'),
  updateSelection: createAction('updateSelection'),
});
