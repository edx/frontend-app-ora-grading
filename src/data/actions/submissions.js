import { StrictDict } from 'utils';
import { createActionFactory } from './utils';

export const dataKey = 'submissions';
const createAction = createActionFactory(dataKey);

/**
 * Load the basic list-level submission data, keyed by submission id
 * @param {obj} submissionListData
 */
const loadList = createAction('loadList');

export default StrictDict({
  loadList,
});
