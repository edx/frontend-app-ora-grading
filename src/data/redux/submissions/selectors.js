import _ from 'lodash';
import { createSelector } from 'reselect';

import { StrictDict } from 'utils';
import { lockStatuses } from 'data/services/lms/constants';

export const simpleSelectors = {
  allSubmissions: state => state.submissions.allSubmissions,
};

/**
 * Returns the submission list in default order for the table.
 */
export const listData = createSelector(
  [simpleSelectors.allSubmissions],
  (allSubmissions) => {
    const submissionIds = Object.keys(allSubmissions);
    const submissionList = submissionIds.map(id => {
      const { gradeStatus, lockStatus, ...rest } = allSubmissions[id];
      const gradingStatus = (lockStatus === lockStatuses.unlocked ? gradeStatus : lockStatus);
      return { gradingStatus, ...rest };
    });
    return _.sortBy(
      submissionList,
      ['submissionDate'],
    );
  },
);

export default StrictDict({
  ...simpleSelectors,
  listData,
});
