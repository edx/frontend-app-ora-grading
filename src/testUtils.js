// eslint-disable-next-line import/prefer-default-export
export const formatMessage = (msg, values) => {
  let message = msg.defaultMessage;
  if (values === undefined) { return message; }
  Object.keys(values).forEach(key => {
    // eslint-disable-next-line
    message = message.replace(`{${key}}`, values[key]);
  });
  return message;
};
