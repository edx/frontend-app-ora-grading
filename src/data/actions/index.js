import { StrictDict } from 'utils';

import app from './app';
import grading from './grading';
import submissions from './submissions';

export default StrictDict({
  app,
  grading,
  submissions,
});
