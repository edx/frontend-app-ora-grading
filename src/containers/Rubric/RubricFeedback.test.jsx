import React from 'react';
import { shallow } from 'enzyme';

import actions from 'data/actions';
import selectors from 'data/selectors';
import {
  feedbackRequirement,
  gradeStatuses,
} from 'data/services/lms/constants';

import { formatMessage } from 'testUtils';

import {
  RubricFeedback,
  mapDispatchToProps,
  mapStateToProps,
} from './RubricFeedback';

jest.mock('components/InfoPopover', () => 'InfoPopover');

jest.mock('@edx/paragon', () => {
  const Form = () => 'Form';
  Object.defineProperty(Form, 'name', { value: 'Form' });
  Form.Group = 'Form.Group';
  Form.Label = 'Form.Label';
  Form.Control = 'Form.Control';
  return { Form };
});

jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      isGrading: jest.fn((...args) => ({ isGragrding: args })),
      rubric: {
        feedbackConfig: jest.fn((...args) => ({
          rubricFeedbackConfig: args,
        })),
        feedbackPrompt: jest.fn((...args) => ({
          rubricFeedbackPrompt: args,
        })),
      },
    },
    grading: {
      selected: {
        overallFeedback: jest.fn((...args) => ({
          selectedOverallFeedback: args,
        })),
      },
    },
  },
}));

describe('Rubric Feedback component', () => {
  const props = {
    intl: { formatMessage },
    config: 'config stirng',
    isGrading: true,
    value: 'some value',
    feedbackPrompt: 'feedback prompt',
    gradeStatus: gradeStatuses.ungraded,
    setValue: jest.fn().mockName('this.props.setValue'),
  };

  let el;
  beforeEach(() => {
    el = shallow(<RubricFeedback {...props} />);
    el.instance().onChange = jest.fn().mockName('this.onChange');
  });
  describe('snapshot', () => {
    test('is grading', () => {
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('is graded', () => {
      el.setProps({
        isGrading: false,
        gradeStatus: gradeStatuses.graded,
      });
      expect(el.instance().render()).toMatchSnapshot();
    });

    test('is configure to disabled', () => {
      el.setProps({
        config: feedbackRequirement.disabled,
      });
      expect(el.instance().render()).toMatchSnapshot();
    });
  });

  describe('component', () => {
    describe('render', () => {
      test('is grading (everything show up and the input is editable)', () => {
        expect(el.isEmptyRender()).toEqual(false);
        const input = el.find('.rubric-feedback.feedback-input');
        expect(input.prop('disabled')).toEqual(false);
        expect(input.prop('value')).toEqual(props.value);
      });

      test('is graded (the input are disabled)', () => {
        el.setProps({
          isGrading: false,
          gradeStatus: gradeStatuses.graded,
        });
        expect(el.isEmptyRender()).toEqual(false);
        const input = el.find('.rubric-feedback.feedback-input');
        expect(input.prop('disabled')).toEqual(true);
        expect(input.prop('value')).toEqual(props.value);
      });
      test('is configure to disabled (this input does not get render)', () => {
        el.setProps({
          config: feedbackRequirement.disabled,
        });
        expect(el.isEmptyRender()).toEqual(true);
      });
    });
    describe('behavior', () => {
      test('onChange set value', () => {
        el = shallow(<RubricFeedback {...props} />);
        el.instance().onChange({
          target: {
            value: 'some value',
          },
        });
        expect(props.setValue).toBeCalledTimes(1);
      });
    });
  });

  describe('mapStateToProps', () => {
    const testState = { abitaryState: 'some data' };
    let mapped;
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('selectors.app.isGrading', () => {
      expect(mapped.isGrading).toEqual(selectors.app.isGrading(testState));
    });

    test('selectors.app.rubricFeedbackConfig', () => {
      expect(mapped.config).toEqual(
        selectors.app.rubric.feedbackConfig(testState),
      );
    });

    test('selectors.grading.selected.overallFeedback', () => {
      expect(mapped.value).toEqual(
        selectors.grading.selected.overallFeedback(testState),
      );
    });

    test('selectors.app.rubric.feedbackPrompt', () => {
      expect(mapped.feedbackPrompt).toEqual(
        selectors.app.rubric.feedbackPrompt(testState),
      );
    });
  });

  describe('mapDispatchToProps', () => {
    test('maps actions.grading.setRubricFeedback to setValue prop', () => {
      expect(mapDispatchToProps.setValue).toEqual(
        actions.grading.setRubricFeedback,
      );
    });
  });
});
