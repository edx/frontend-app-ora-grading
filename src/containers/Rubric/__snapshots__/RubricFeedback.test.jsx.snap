// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rubric Feedback component snapshot is configure to disabled 1`] = `null`;

exports[`Rubric Feedback component snapshot is graded 1`] = `
<Form.Group>
  <Form.Label
    className="criteria-label"
  >
    <span
      className="criteria-title"
    >
      <FormattedMessage
        defaultMessage="Overall comments"
        description="Rubric overall commnents label"
        id="ora-grading.Rubric.overallComments"
      />
    </span>
    <InfoPopover>
      <div>
        feedback prompt
      </div>
    </InfoPopover>
  </Form.Label>
  <Form.Control
    as="input"
    className="rubric-feedback feedback-input"
    disabled={true}
    floatingLabel="Comments"
    onChange={[MockFunction this.onChange]}
    value="some value"
  />
</Form.Group>
`;

exports[`Rubric Feedback component snapshot is grading 1`] = `
<Form.Group>
  <Form.Label
    className="criteria-label"
  >
    <span
      className="criteria-title"
    >
      <FormattedMessage
        defaultMessage="Overall comments"
        description="Rubric overall commnents label"
        id="ora-grading.Rubric.overallComments"
      />
    </span>
    <InfoPopover>
      <div>
        feedback prompt
      </div>
    </InfoPopover>
  </Form.Label>
  <Form.Control
    as="input"
    className="rubric-feedback feedback-input"
    disabled={false}
    floatingLabel="Add comments"
    onChange={[MockFunction this.onChange]}
    value="some value"
  />
</Form.Group>
`;
